src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,263 @@
+            //update the swapped operator's operatorIdToIndexHistory list with a new entry, as their index has now changed
+            _updateOperatorIdToIndexHistory(operatorIdToSwap, quorumNumber, indexOfOperatorToRemove);
+        } 
+        // marking the final entry in the deregistering operator's operatorIdToIndexHistory entry with the deregistration block number, 
+        // setting the index to OPERATOR_DEREGISTERED_INDEX. Note that this is a special meaning, and any other index value represents a real index
+        _updateOperatorIdToIndexHistory(operatorId, quorumNumber, OPERATOR_DEREGISTERED_INDEX);
+    }
+
+
+    /**
+     * @notice Returns the total number of operators of the service for the given `quorumNumber` at the given `blockNumber`
+     * @dev Returns zero if the @param blockNumber is from before the @param quorumNumber existed, and returns the current number of total operators if the @param blockNumber is in the future.
+     */
+    function _getTotalOperatorsForQuorumAtBlockNumber(uint8 quorumNumber, uint32 blockNumber) internal view returns (uint32){
+        // store list length in memory
+        uint256 totalOperatorsHistoryLength = _totalOperatorsHistory[quorumNumber].length;
+        // if there are no entries in the total operator history, return 0
+        if (totalOperatorsHistoryLength == 0) {
+            return 0;
+        }
+
+        // if `blockNumber` is from before the `quorumNumber` existed, return `0`
+        if (blockNumber < _totalOperatorsHistory[quorumNumber][0].fromBlockNumber) {
+            return 0;
+        }
+
+        // loop backwards through the total operator history to find the total number of operators at the given block number
+        for (uint256 i = 0; i <= totalOperatorsHistoryLength - 1; i++) {
+            uint256 listIndex = (totalOperatorsHistoryLength - 1) - i;
+            OperatorIndexUpdate memory totalOperatorUpdate = _totalOperatorsHistory[quorumNumber][listIndex];
+            // look for the first update that began before or at `blockNumber`
+            if (totalOperatorUpdate.fromBlockNumber <= blockNumber) {
+                return _totalOperatorsHistory[quorumNumber][listIndex].index;
+            }
+        }        
+        return _totalOperatorsHistory[quorumNumber][0].index;
+    }
+    
+
+    /// @notice Returns the index of the `operatorId` at the given `blockNumber` for the given `quorumNumber`, or `OPERATOR_DEREGISTERED_INDEX` if the operator was not registered for the `quorumNumber` at blockNumber
+    function _getIndexOfOperatorForQuorumAtBlockNumber(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber) internal view returns(uint32) {
+        uint256 operatorIndexHistoryLength = _operatorIdToIndexHistory[operatorId][quorumNumber].length;
+        // loop backward through index history to find the index of the operator at the given block number
+        for (uint i = 0; i < _operatorIdToIndexHistory[operatorId][quorumNumber].length; i++) {
```

Comment:
we've just stored `_operatorIdToIndexHistory[operatorId][quorumNumber].length` in memory and should use it in the loop condition.
also style nit: generally prefer 'uint256' over the alias 'uint'